sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala

```diff
@@ -1432,6 +1681,53 @@ case class Logarithm(left: Expression, right: Expression)
     newLeft: Expression, newRight: Expression): Logarithm = copy(left = newLeft, right = newRight)
 }
 
+trait BaseBinaryExpression extends BinaryExpression
+  with ExpectsInputTypes
+  with Serializable
+  with ImplicitCastInputTypes {
+  val child: Expression
+  val scale: Expression
+  override def left: Expression = child
+  override def right: Expression = scale
+  override def nullable: Boolean = true
+  override def foldable: Boolean = child.foldable
+
+  /**
+   * Expected input types from child expressions. The i-th position in the returned seq indicates
+   * the type requirement for the i-th child.
+   *
+   * The possible values at each position are:
+   *   1. a specific data type, such as LongType or StringType.
+   *   2. a non-leaf abstract data type,
+   *      such as NumericType, IntegralType, FractionalType.
+   */
+  override def inputTypes: Seq[AbstractDataType] = Seq(NumericType, IntegerType)
+
+  // Avoid repeated evaluation since `scale` is a constant int,
+  // avoid unnecessary `child` evaluation in both codegen and non-codegen eval
+  // by checking if scaleV == null as well.
+  protected lazy val scaleV: Any = scale.eval(EmptyRow)
+
+  protected lazy val scaleValue: Int = scaleV.asInstanceOf[Int]
+
+  override def checkInputDataTypes(): TypeCheckResult = {
```

Comment:
Does this mean we force the second argument to be a constant? Why?

Reply:
other related  udfs `ceil` , `floor` also  does not work with dynamic values.
To make it consistent with those udfs,  here `scale`is made constant . This part of the code was moved from  `class RoundBase ` to new trait `BaseBinaryExpression`